### PR TITLE
Added an Appendix section describing some message flows.

### PR DESCRIPTION
**Pull request by ctjhai**
The following three flows are considered:
* AKE is not used, peers selects NONE on all cases;
* AKE is not used in IKE_SA_INIT, but used in CREATE_CHILD_SA; and
* AKE negotiation results in NO_PROPOSAL_CHOSEN.

### CODE REVIEW

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml [hunk 1]
```diff
@@ -885,6 +885,129 @@ Type     Description                   Used In
...
+
+HDR(IKE_AUTH), SK{ IDi, AUTH, SAi2, TSi, TSr } --->
+                         <--- HDR(IKE_AUTH), SK{ IDr, AUTH, SAr2,
+                              TSi, TSr }
+      ]]></artwork></figure>
+    </section>
+
+    <section title="Additional Key Exchange for CHILD SA only" anchor="sample-exchanges-ake-child-sas">
+      <t>The exchanges below show that the initiator does not propose the use of additional key exchanges
+      to establish an IKE SA, but they are required in order to establish a CHILD SA.  Note that the
```
Comment (smyslov) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: s/CHILD SA/Child SA
Note, that the very first Child SA will be created as result of IKE_AUTH, so if both the initiator and the responder have policy to create only quantum-protected IPsec SA, then both should exchange CHILDLESS_IKEV2_SUPPORTED in the IKE_SA_INIT and don't create ESP SA in IKE_AUTH.

A few words should be added that IKE SA is then immediately rekeyed and then Child SAs are creared as usual.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml [hunk 2]
```diff
@@ -885,6 +885,129 @@ Type     Description                   Used In
...
+      initiator does not advertise N(INTERMEDIATE_EXCHANGE_SUPPORTED) in the IKE_SA_INIT request.</t>
+
+      <figure><artwork align="center"><![CDATA[
+Initiator                     Responder
+------------------------------------------------------------------------
+HDR(IKE_SA_INIT), SAi1(.. AKE*...), --->
+KEi1, Ni, N(IKEV2_FRAG_SUPPORTED)
+                         <--- HDR(IKE_SA_INIT), SAr1(.. AKE*...),
+                              KEr1, Nr, N(IKEV2_FRAG_SUPPORTED)
+
```
Comment (smyslov) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: Should we also remove AKE* from SA payload in this case?

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml [hunk 3]
```diff
@@ -885,6 +885,129 @@ Type     Description                   Used In
...
+                                  Transform AKE1 (ID = NONE)
+                                  Transform AKE2 (ID = NONE)
+
+HDR(IKE_AUTH), SK{ IDi, AUTH, SAi2, TSi, TSr } --->
+                         <--- HDR(IKE_AUTH), SK{ IDr, AUTH, SAr2,
+                              TSi, TSr }
+      ]]></artwork></figure>
+    </section>
+
+    <section title="Additional Key Exchange for CHILD SA only" anchor="sample-exchanges-ake-child-sas">
```
Comment (smyslov) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: I think that it's better to rename the section to "Additional Key Exchange in the CREATE_CHILD_SA Exchange Only". And I think it's better to show how to create fully QC-protected IKE SA, in this case so a clarification should be added that we rekey IKE SA (and not Child SA that was created in IKE_AUTH).

**Review by smyslov — CHANGES_REQUESTED** [review]

**Comment by smyslov** [comment]
HI CJ, thank you for this diagrams. I wrote some comments. I may directly edit the file if you agree.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml (hunk 2 above)
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: Ah good catch, thanks. Yes, it should be removed.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml (hunk 3 above)
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: Happy with that.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml (hunk 1 above)
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:936]: > HI CJ, thank you for this diagrams. I wrote some comments. I may directly edit the file if you agree.

Hi @smyslov, many thanks for your review. Please go ahead and edit the file directly :-)

**Comment by ctjhai** [comment]
> HI CJ, thank you for this diagrams. I wrote some comments. I may directly edit the file if you agree.

Hi @smyslov, many thanks for your review. Please go ahead and edit the file directly :-)

**Comment by smyslov** [comment]
Ah, I was thinking that I'm able to directly edit your proposed changes, but it seems that github doesn't allow me to do it (or I don't know how to make this possible) :-( Sorry.

**Comment by ctjhai** [comment]
> Ah, I was thinking that I'm able to directly edit your proposed changes, but it seems that github doesn't allow me to do it (or I don't know how to make this possible) :-( Sorry.

Not to worry, I have amended some of the text. I will share the DIFF with you directly over email.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml [hunk 4]
```diff
@@ -750,6 +751,18 @@ KEYMAT = prf+ (SK_d, KE  | Ni | Nr | KE(1) |  ... KE(n))
...
+            <section title="Interaction with Childless IKE SA">
+            <t>It is also possible to establish a fully quantum-resistant IKE SAs from multiple key exchanges
+            without using IKE_INTERMEDIATE exchanges.  In this case, both peers should immediately rekey the IKE SA
+            and prevent the very first Child SA from being created, following the IKE_AUTH exchange.  In order to
+            achieve this, both peers MUST support the experimental <xref target="RFC6023"/> and exchange
+            CHILDLESS_IKEV2_SUPPORTED notification in the IKE_SA_INIT exchange.  The rekey of the IKE SA is done using
+            CREATE_CHILD_SA exchange where the multiple key exchange is negotiated and the key exchange payload is carried
+            in the IKE_FOLLOWUP_KE message.
+            </t>
+            </section>
```
Comment (smyslov) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:773]: I would add one more para here:

<t> Note, that if initial IKE SA is used to transfer sensitive information, then this information will not be protected using additional (e.g. quantum safe) key exchanges, so this scenario may be inappropriate. One of examples when sensitive information is transferred in initial IKE SA is G-IKEv2 protocol <xref target=xxx />.
</t>

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml [hunk 5]
```diff
@@ -750,6 +751,18 @@ KEYMAT = prf+ (SK_d, KE  | Ni | Nr | KE(1) |  ... KE(n))
                 </t>
 
             </section>
+
+            <section title="Interaction with Childless IKE SA">
+            <t>It is also possible to establish a fully quantum-resistant IKE SAs from multiple key exchanges
+            without using IKE_INTERMEDIATE exchanges.  In this case, both peers should immediately rekey the IKE SA
+            and prevent the very first Child SA from being created, following the IKE_AUTH exchange.  In order to
+            achieve this, both peers MUST support the experimental <xref target="RFC6023"/> and exchange
```
Comment (smyslov) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: I would avoid "MUST" here. E.g. if we mean G-IKEv2, then there is no need for this "MUST". I also think that it's a local policy which dictates how Child SAs must be protected. I'd rather rephrase alone the lines: 

If peers' local policy is that all Child SAs should be fully protected, then peers can avoid creating the very first Child SA by adopting <xref target="RFC6023"/>. In this case peers exchange CHILDLESS_IKEV2_SUPPORTED notification in the IKE_SA_INIT exchange.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml [hunk 6]
```diff
@@ -885,6 +899,133 @@ Type     Description                   Used In
...
+HDR(IKE_AUTH), SK{ IDi, AUTH, SAi2, TSi, TSr } --->
+                         <--- HDR(IKE_AUTH), SK{ IDr, AUTH, SAr2,
+                              TSi, TSr }
+      ]]></artwork></figure>
+    </section>
+
+    <section title="Additional Key Exchange in the CREATE_CHILD_SA Exchange only" anchor="sample-exchanges-ake-child-sas">
+      <t>The exchanges below show that the initiator does not propose the use of additional key exchanges
+      to establish an IKE SA, but they are required in order to establish a Child SA.  Note that the
+      initiator does not advertise N(INTERMEDIATE_EXCHANGE_SUPPORTED) in the IKE_SA_INIT request. In order
```
Comment (smyslov) [inline draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: I'd delete this sentence. Or added some clarification. Note, that IKE_INTERMEDIATE may be negotiated in this case for purposes other than multiple key exchanges. The main reason peers agree not to do multiple key exchanges in initial IKE SA is that the initiator doesn't include any AKE transforms in SA (or included them with only NONE values) and the responder found this appropriate. Support for IKE_INTERMEDIATE may or may not be negotiated in this case, it is not relevant.

**Review by smyslov — CHANGES_REQUESTED** [review]

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml (hunk 6 above)
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: Yes, you are right, I forgot about that. I have removed this sentence in the latest update.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml (hunk 4 above)
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml:773]: Thanks, I've modified the paragraph on this.

File: draft-ietf-ipsecme-ikev2-multiple-ke-00.xml (hunk 5 above)
Comment (ctjhai) [reply draft-ietf-ipsecme-ikev2-multiple-ke-00.xml]: Yep, the paragraph has been updated.

**Review by smyslov — APPROVED** [review]
I suggest to add a reference to G-IKEv2 draft, since it's not well known protocol. Otherwise I think that we should merge this PR and my PR too, (unless folks have more comments and opinions) and then review again the resulting text. We have already quite a lot changes, so it'll be easier to look at the final text than at diffs.

**Comment by ctjhai** [comment]
> I suggest to add a reference to G-IKEv2 draft, since it's not well known protocol. Otherwise I think that we should merge this PR and my PR too, (unless folks have more comments and opinions) and then review again the resulting text. We have already quite a lot changes, so it'll be easier to look at the final text than at diffs.

Yep, it's now added. Agreed, we should merge this. It would be easier to review the entire document as a whole. I'll email folks about the merged PRs so that they can review the updated document, if they like.

**Pull request merged**